File: rating_engine.py
"""
Enterprise Credit Rating Engine
8-dimension scoring system (0-100) with letter grades (AAA-CCC)
Reference: Moody's/S&P credit rating methodology + Morningstar equity research
"""

import json
import datetime as dt
from collections import defaultdict


def _score_scale(val, low, mid, high, reverse=False):
    """Map a value to 0-25 score scale with interpolation.
    low=7.5pts, mid=15pts, high=25pts. reverse=True for lower-is-better metrics."""
    if val is None:
        return 0
    if reverse:
        val = -val
        low, mid, high = -low, -mid, -high
    if val >= high:
        return 25
    elif val >= mid:
        t = (val - mid) / max(high - mid, 1e-9)
        return 15 + t * 10
    elif val >= low:
        t = (val - low) / max(mid - low, 1e-9)
        return 7.5 + t * 7.5
    else:
        t = val / max(low, 1e-9) if low > 0 else 0
        return max(0, t * 7.5)


def compute_enterprise_rating(
    net_margin=None,
    gross_margin=None,
    roe=None,
    roa=None,
    debt_ratio=None,
    current_ratio=None,
    interest_coverage=None,
    asset_turnover=None,
    inv_turnover=None,
    recv_turnover=None,
    revenue_growth=None,
    profit_growth=None,
    pe_ratio=None,
    pb_ratio=None,
    ps_ratio=None,
    industry_pe=None,
    industry_pb=None,
    industry_ps=None,
    operating_cash_flow=None,
    net_profit=None,
    free_cash_flow=None,
    gross_margin_3y=None,
    market_cap_rank=None,
    industry_avg_debt=None,
    industry_avg_roe=None,
    industry_avg_margin=None,
):
    """8-dimension enterprise rating. Returns dict with total score, grade, and details."""

    dims = {}

    # === DIMENSION 1: Profitability (20%) ===
    p_scores = {}
    p_scores["net_margin"] = _score_scale(net_margin, 3, 10, 25)
    p_scores["gross_margin"] = _score_scale(gross_margin, 20, 40, 70)
    if net_margin is not None and gross_margin is not None:
        expense = gross_margin - net_margin
        p_scores["expense_control"] = _score_scale(expense, 50, 25, 10, reverse=True)
    else:
        p_scores["expense_control"] = 0
    p_avg = sum(p_scores.values()) / len(p_scores) if p_scores else 0
    dims["profitability"] = {
        "score": round(p_avg, 1),
        "max": 25,
        "weight": 0.20,
        "weighted": round(p_avg * 0.20, 2),
        "details": {k: round(v, 1) for k, v in p_scores.items()},
        "label": "盈利能力",
    }

    # === DIMENSION 2: Capital Efficiency (15%) ===
    c_scores = {}
    c_scores["roe"] = _score_scale(roe, 5, 15, 30)
    c_scores["roa"] = _score_scale(roa, 2, 8, 18)
    if roe is not None and roa is not None and roe > 0:
        leverage_quality = roa / max(roe, 1e-9)
        c_scores["leverage_quality"] = _score_scale(leverage_quality, 0.15, 0.4, 0.7)
    else:
        c_scores["leverage_quality"] = 0
    c_avg = sum(c_scores.values()) / len(c_scores) if c_scores else 0
    dims["capital_efficiency"] = {
        "score": round(c_avg, 1),
        "max": 25,
        "weight": 0.15,
        "weighted": round(c_avg * 0.15, 2),
        "label": "资本效率",
    }

    # === DIMENSION 3: Financial Safety (15%) ===
    f_scores = {}
    f_scores["debt_ratio"] = _score_scale(debt_ratio, 70, 50, 30, reverse=True)
    f_scores["current_ratio"] = _score_scale(current_ratio, 0.8, 1.5, 3.0)
    if interest_coverage is not None:
        f_scores["interest_coverage"] = _score_scale(interest_coverage, 1.5, 4, 10)
    else:
        f_scores["interest_coverage"] = 0
    f_avg = sum(f_scores.values()) / len(f_scores) if f_scores else 0
    dims["financial_safety"] = {
        "score": round(f_avg, 1),
        "max": 25,
        "weight": 0.15,
        "weighted": round(f_avg * 0.15, 2),
        "label": "财务安全",
    }

    # === DIMENSION 4: Operating Efficiency (10%) ===
    o_scores = {}
    o_scores["asset_turnover"] = _score_scale(asset_turnover, 0.3, 0.8, 2.0)
    o_scores["inv_turnover"] = _score_scale(inv_turnover, 2, 6, 15)
    o_scores["recv_turnover"] = _score_scale(recv_turnover, 3, 8, 20)
    o_avg = sum(o_scores.values()) / len(o_scores) if o_scores else 0
    dims["operating_efficiency"] = {
        "score": round(o_avg, 1),
        "max": 25,
        "weight": 0.10,
        "weighted": round(o_avg * 0.10, 2),
        "label": "营运效率",
    }

    # === DIMENSION 5: Growth (15%) ===
    g_scores = {}
    g_scores["revenue_growth"] = _score_scale(revenue_growth, -5, 10, 30)
    g_scores["profit_growth"] = _score_scale(profit_growth, -10, 15, 40)
    g_avg = sum(g_scores.values()) / len(g_scores) if g_scores else 0
    dims["growth"] = {
        "score": round(g_avg, 1),
        "max": 25,
        "weight": 0.15,
        "weighted": round(g_avg * 0.15, 2),
        "label": "成长性",
    }

    # === DIMENSION 6: Valuation (10%) ===
    v_scores = {}
    if pe_ratio is not None and industry_pe is not None and industry_pe > 0:
        pe_relative = pe_ratio / industry_pe
        v_scores["pe_relative"] = _score_scale(pe_relative, 2.0, 1.2, 0.6, reverse=True)
    elif pe_ratio is not None:
        v_scores["pe_absolute"] = _score_scale(pe_ratio, 50, 20, 8, reverse=True)
    else:
        v_scores["pe"] = 0

    if pb_ratio is not None and industry_pb is not None and industry_pb > 0:
        pb_relative = pb_ratio / industry_pb
        v_scores["pb_relative"] = _score_scale(pb_relative, 2.5, 1.3, 0.5, reverse=True)
    elif pb_ratio is not None:
        v_scores["pb_absolute"] = _score_scale(pb_ratio, 5, 2, 0.8, reverse=True)

    if ps_ratio is not None and industry_ps is not None and industry_ps > 0:
        ps_relative = ps_ratio / industry_ps
        v_scores["ps_relative"] = _score_scale(ps_relative, 2.5, 1.2, 0.5, reverse=True)
    elif ps_ratio is not None:
        v_scores["ps_absolute"] = _score_scale(ps_ratio, 8, 3, 0.5, reverse=True)

    v_avg = sum(v_scores.values()) / len(v_scores) if v_scores else 0
    dims["valuation"] = {
        "score": round(v_avg, 1),
        "max": 25,
        "weight": 0.10,
        "weighted": round(v_avg * 0.10, 2),
        "label": "估值水平",
    }

    # === DIMENSION 7: Cash Flow Quality (10%) ===
    cf_scores = {}
    if operating_cash_flow is not None and net_profit is not None and abs(net_profit) > 1e-9:
        cfo_np_ratio = operating_cash_flow / abs(net_profit)
        cf_scores["cfo_quality"] = _score_scale(cfo_np_ratio, 0.5, 1.0, 1.5)
    else:
        cf_scores["cfo_quality"] = 0

    if free_cash_flow is not None:
        cf_scores["fcf"] = _score_scale(free_cash_flow, -0.05, 0.02, 0.08)
    else:
        cf_scores["fcf"] = 0

    cf_avg = sum(cf_scores.values()) / len(cf_scores) if cf_scores else 0
    dims["cash_flow"] = {
        "score": round(cf_avg, 1),
        "max": 25,
        "weight": 0.10,
        "weighted": round(cf_avg * 0.10, 2),
        "label": "现金流质量",
    }

    # === DIMENSION 8: Moat (5%) ===
    m_scores = {}
    if gross_margin_3y is not None and len(gross_margin_3y) >= 3:
        gm_std = _std(gross_margin_3y)
        m_scores["margin_stability"] = _score_scale(gm_std, 8, 3, 1, reverse=True)
    elif gross_margin is not None:
        m_scores["margin_stability"] = 12.5
    else:
        m_scores["margin_stability"] = 0

    if market_cap_rank is not None:
        m_scores["market_position"] = _score_scale(market_cap_rank, 50, 20, 5, reverse=True)
    elif gross_margin is not None and gross_margin > 50:
        m_scores["market_position"] = 18.0
    else:
        m_scores["market_position"] = 5.0

    m_avg = sum(m_scores.values()) / len(m_scores) if m_scores else 0
    dims["moat"] = {
        "score": round(m_avg, 1),
        "max": 25,
        "weight": 0.05,
        "weighted": round(m_avg * 0.05, 2),
        "label": "护城河",
    }

    # === AGGREGATE ===
    total = sum(d["weighted"] for d in dims.values())
    total_pct = total / 25.0 * 100  # normalize to 0-100

    # Letter grade (Moody's-style)
    if total_pct >= 90:
        grade = "AAA"
    elif total_pct >= 80:
        grade = "AA"
    elif total_pct >= 70:
        grade = "A"
    elif total_pct >= 60:
        grade = "BBB"
    elif total_pct >= 50:
        grade = "BB"
    elif total_pct >= 35:
        grade = "B"
    else:
        grade = "CCC"

    # Investment recommendation
    if total_pct >= 75:
        recommendation = "优质标的，估值合理时可积极配置"
    elif total_pct >= 60:
        recommendation = "基本面稳健，关注估值安全边际后可配置"
    elif total_pct >= 45:
        recommendation = "基本面中等，需等待更多积极信号"
    elif total_pct >= 30:
        recommendation = "基本面偏弱，谨慎观望"
    else:
        recommendation = "财务风险较高，建议回避"

    # Risk flags
    risks = []
    if debt_ratio is not None and debt_ratio > 70:
        risks.append("高负债率(>70%)")
    if current_ratio is not None and current_ratio < 1.0:
        risks.append("流动比率不足(<1.0)")
    if net_margin is not None and net_margin < 3:
        risks.append("净利率极低(<3%)")
    if roe is not None and roe < 5:
        risks.append("ROE过低(<5%)")
    if revenue_growth is not None and revenue_growth < -10:
        risks.append("营收大幅下滑(>-10%)")

    # Strength flags
    strengths = []
    if roe is not None and roe > 20:
        strengths.append("ROE优秀(>20%)")
    if gross_margin is not None and gross_margin > 50:
        strengths.append("强定价权(毛利率>50%)")
    if net_margin is not None and net_margin > 20:
        strengths.append("高利润率(>20%)")
    if debt_ratio is not None and debt_ratio < 30:
        strengths.append("极低负债(<30%)")
    if revenue_growth is not None and revenue_growth > 20:
        strengths.append("高增长(>20%)")

    # Dimension summary for quick display
    dim_summary = {}
    for k, d in dims.items():
        pct = d["score"] / d["max"] * 100 if d["max"] else 0
        if pct >= 80:
            flag = "强"
        elif pct >= 60:
            flag = "良"
        elif pct >= 40:
            flag = "中"
        else:
            flag = "弱"
        dim_summary[d["label"]] = {"score": round(d["score"], 1), "pct": round(pct, 0), "flag": flag, "weight": d["weight"]}

    return {
        "total_score": round(total_pct, 1),
        "grade": grade,
        "recommendation": recommendation,
        "dimensions": dims,
        "dim_summary": dim_summary,
        "risks": risks,
        "strengths": strengths,
    }


def _std(values):
    if not values or len(values) < 2:
        return 0
    n = len(values)
    mean = sum(values) / n
    variance = sum((x - mean) ** 2 for x in values) / (n - 1)
    return variance ** 0.5

File: frontend/src/services/ratingEngine.ts
/**
 * Enterprise Credit Rating Engine - 8-dimension scoring system
 * Translated from Python rating_engine.py
 * Reference: Moody's/S&P credit rating methodology + Morningstar equity research
 */

interface DimResult {
  score: number;
  max: number;
  weight: number;
  weighted: number;
  label: string;
  details: Record<string, number>;
}

interface RatingResult {
  total_score: number;
  grade: string;
  recommendation: string;
  dimensions: Record<string, DimResult>;
  dim_summary: Record<string, { score: number; pct: number; flag: string; weight: number }>;
  risks: string[];
  strengths: string[];
}

function _scoreScale(
  val: number | null | undefined,
  low: number,
  mid: number,
  high: number,
  reverse = false
): number {
  if (val == null) return 0;
  let v = val;
  let lo = low,
    mi = mid,
    hi = high;
  if (reverse) {
    v = -v;
    lo = -lo;
    mi = -mi;
    hi = -hi;
  }
  if (v >= hi) return 25;
  if (v >= mi) {
    const t = (v - mi) / Math.max(hi - mi, 1e-9);
    return 15 + t * 10;
  }
  if (v >= lo) {
    const t = (v - lo) / Math.max(mi - lo, 1e-9);
    return 7.5 + t * 7.5;
  }
  const t = lo > 0 ? v / lo : 0;
  return Math.max(0, t * 7.5);
}

function avg(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

export interface RatingInput {
  net_margin?: number | null;
  gross_margin?: number | null;
  roe?: number | null;
  roa?: number | null;
  debt_ratio?: number | null;
  current_ratio?: number | null;
  asset_turnover?: number | null;
  inventory_turnover?: number | null;
  receivable_turnover?: number | null;
  revenue_growth?: number | null;
  profit_growth?: number | null;
  pe_ratio?: number | null;
  operating_cash_flow?: number | null;
  net_profit?: number | null;
  total_assets?: number | null;
}

export function computeEnterpriseRating(inp: RatingInput): RatingResult {
  const dims: Record<string, DimResult> = {};

  // === DIM 1: Profitability (20%) ===
  const pScores: Record<string, number> = {};
  pScores["net_margin"] = _scoreScale(inp.net_margin, 3, 10, 25);
  pScores["gross_margin"] = _scoreScale(inp.gross_margin, 20, 40, 70);
  if (inp.net_margin != null && inp.gross_margin != null) {
    pScores["expense_control"] = _scoreScale(
      inp.gross_margin - inp.net_margin,
      50,
      25,
      10,
      true
    );
  } else {
    pScores["expense_control"] = 0;
  }
  const pAvg = avg(Object.values(pScores));
  dims["profitability"] = {
    score: Math.round(pAvg * 10) / 10,
    max: 25,
    weight: 0.2,
    weighted: Math.round(pAvg * 0.2 * 100) / 100,
    label: "盈利能力",
    details: Object.fromEntries(
      Object.entries(pScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 2: Capital Efficiency (15%) ===
  const cScores: Record<string, number> = {};
  cScores["roe"] = _scoreScale(inp.roe, 5, 15, 30);
  cScores["roa"] = _scoreScale(inp.roa, 2, 8, 18);
  if (inp.roe != null && inp.roa != null && inp.roe > 0) {
    cScores["leverage_quality"] = _scoreScale(inp.roa / inp.roe, 0.15, 0.4, 0.7);
  } else {
    cScores["leverage_quality"] = 0;
  }
  const cAvg = avg(Object.values(cScores));
  dims["capital_efficiency"] = {
    score: Math.round(cAvg * 10) / 10,
    max: 25,
    weight: 0.15,
    weighted: Math.round(cAvg * 0.15 * 100) / 100,
    label: "资本效率",
    details: Object.fromEntries(
      Object.entries(cScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 3: Financial Safety (15%) ===
  const fScores: Record<string, number> = {};
  fScores["debt_ratio"] = _scoreScale(inp.debt_ratio, 70, 50, 30, true);
  fScores["current_ratio"] = _scoreScale(inp.current_ratio, 0.8, 1.5, 3.0);
  fScores["interest_coverage"] = 0;
  const fAvg = avg(Object.values(fScores));
  dims["financial_safety"] = {
    score: Math.round(fAvg * 10) / 10,
    max: 25,
    weight: 0.15,
    weighted: Math.round(fAvg * 0.15 * 100) / 100,
    label: "财务安全",
    details: Object.fromEntries(
      Object.entries(fScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 4: Operating Efficiency (10%) ===
  const oScores: Record<string, number> = {};
  oScores["asset_turnover"] = _scoreScale(inp.asset_turnover, 0.3, 0.8, 2.0);
  oScores["inv_turnover"] = _scoreScale(inp.inventory_turnover, 2, 6, 15);
  oScores["recv_turnover"] = _scoreScale(inp.receivable_turnover, 3, 8, 20);
  const oAvg = avg(Object.values(oScores));
  dims["operating_efficiency"] = {
    score: Math.round(oAvg * 10) / 10,
    max: 25,
    weight: 0.1,
    weighted: Math.round(oAvg * 0.1 * 100) / 100,
    label: "营运效率",
    details: Object.fromEntries(
      Object.entries(oScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 5: Growth (15%) ===
  const gScores: Record<string, number> = {};
  gScores["revenue_growth"] = _scoreScale(inp.revenue_growth, -5, 10, 30);
  gScores["profit_growth"] = _scoreScale(inp.profit_growth, -10, 15, 40);
  const gAvg = avg(Object.values(gScores));
  dims["growth"] = {
    score: Math.round(gAvg * 10) / 10,
    max: 25,
    weight: 0.15,
    weighted: Math.round(gAvg * 0.15 * 100) / 100,
    label: "成长性",
    details: Object.fromEntries(
      Object.entries(gScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 6: Valuation (10%) ===
  const vScores: Record<string, number> = {};
  vScores["pe"] = _scoreScale(inp.pe_ratio, 50, 20, 8, true);
  const vAvg = avg(Object.values(vScores));
  dims["valuation"] = {
    score: Math.round(vAvg * 10) / 10,
    max: 25,
    weight: 0.1,
    weighted: Math.round(vAvg * 0.1 * 100) / 100,
    label: "估值水平",
    details: Object.fromEntries(
      Object.entries(vScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 7: Cash Flow Quality (10%) ===
  const cfScores: Record<string, number> = {};
  if (
    inp.operating_cash_flow != null &&
    inp.net_profit != null &&
    Math.abs(inp.net_profit) > 1e-9
  ) {
    cfScores["cfo_quality"] = _scoreScale(
      inp.operating_cash_flow / Math.abs(inp.net_profit),
      0.5,
      1.0,
      1.5
    );
  } else {
    cfScores["cfo_quality"] = 0;
  }
  cfScores["fcf"] = 0;
  const cfAvg = avg(Object.values(cfScores));
  dims["cash_flow"] = {
    score: Math.round(cfAvg * 10) / 10,
    max: 25,
    weight: 0.1,
    weighted: Math.round(cfAvg * 0.1 * 100) / 100,
    label: "现金流质量",
    details: Object.fromEntries(
      Object.entries(cfScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === DIM 8: Moat (5%) ===
  const mScores: Record<string, number> = {};
  if (inp.gross_margin != null) {
    mScores["margin_stability"] = inp.gross_margin >= 60 ? 22 : inp.gross_margin >= 40 ? 18 : inp.gross_margin >= 20 ? 12 : 5;
  } else {
    mScores["margin_stability"] = 0;
  }
  if (inp.gross_margin != null && inp.total_assets != null) {
    mScores["market_position"] =
      inp.gross_margin > 40 && inp.total_assets > 1e9
        ? 22
        : inp.gross_margin > 30 && inp.total_assets > 5e8
          ? 18
          : inp.gross_margin > 20
            ? 12
            : 5;
  } else if (inp.gross_margin != null && inp.gross_margin > 50) {
    mScores["market_position"] = 18;
  } else {
    mScores["market_position"] = 5;
  }
  const mAvg = avg(Object.values(mScores));
  dims["moat"] = {
    score: Math.round(mAvg * 10) / 10,
    max: 25,
    weight: 0.05,
    weighted: Math.round(mAvg * 0.05 * 100) / 100,
    label: "护城河",
    details: Object.fromEntries(
      Object.entries(mScores).map(([k, v]) => [k, Math.round(v * 10) / 10])
    ),
  };

  // === AGGREGATE ===
  const total = Object.values(dims).reduce((s, d) => s + d.weighted, 0);
  const totalPct = (total / 25) * 100;

  let grade: string;
  if (totalPct >= 90) grade = "AAA";
  else if (totalPct >= 80) grade = "AA";
  else if (totalPct >= 70) grade = "A";
  else if (totalPct >= 60) grade = "BBB";
  else if (totalPct >= 50) grade = "BB";
  else if (totalPct >= 35) grade = "B";
  else grade = "CCC";

  let recommendation: string;
  if (totalPct >= 75) recommendation = "优质标的，估值合理时可积极配置";
  else if (totalPct >= 60)
    recommendation = "基本面稳健，关注估值安全边际后可配置";
  else if (totalPct >= 45)
    recommendation = "基本面中等，需等待更多积极信号";
  else if (totalPct >= 30) recommendation = "基本面偏弱，谨慎观望";
  else recommendation = "财务风险较高，建议回避";

  const risks: string[] = [];
  if (inp.debt_ratio != null && inp.debt_ratio > 70) risks.push("高负债率(>70%)");
  if (inp.current_ratio != null && inp.current_ratio < 1.0)
    risks.push("流动比率不足(<1.0)");
  if (inp.net_margin != null && inp.net_margin < 3) risks.push("净利率极低(<3%)");
  if (inp.roe != null && inp.roe < 5) risks.push("ROE过低(<5%)");
  if (inp.revenue_growth != null && inp.revenue_growth < -10)
    risks.push("营收大幅下滑(>-10%)");

  const strengths: string[] = [];
  if (inp.roe != null && inp.roe > 20) strengths.push("ROE优秀(>20%)");
  if (inp.gross_margin != null && inp.gross_margin > 50)
    strengths.push("强定价权(毛利率>50%)");
  if (inp.net_margin != null && inp.net_margin > 20)
    strengths.push("高利润率(>20%)");
  if (inp.debt_ratio != null && inp.debt_ratio < 30)
    strengths.push("极低负债(<30%)");
  if (inp.revenue_growth != null && inp.revenue_growth > 20)
    strengths.push("高增长(>20%)");

  const dimSummary: Record<
    string,
    { score: number; pct: number; flag: string; weight: number }
  > = {};
  for (const [k, d] of Object.entries(dims)) {
    const pct = d.max ? (d.score / d.max) * 100 : 0;
    dimSummary[d.label] = {
      score: Math.round(d.score * 10) / 10,
      pct: Math.round(pct),
      flag: pct >= 80 ? "强" : pct >= 60 ? "良" : pct >= 40 ? "中" : "弱",
      weight: d.weight,
    };
  }

  return {
    total_score: Math.round(totalPct * 10) / 10,
    grade,
    recommendation,
    dimensions: dims,
    dim_summary: dimSummary,
    risks,
    strengths,
  };
}
